Sprint 11 (DispatchQueue, URLRequest, Kingfisher, SwiftKeychainWrapper, ProgressHUD, Race condition, POST-request)

Dassam opened the pull request:
### 1. Fixed race condition in OAuthService. The resulting solution are correctly processed in situations where:
- fetchOAuthToken() was called twice in a row with the same code (the second call does not wait for the first one to complete);
- fetchOAuthToken() was first called with one code, but after some time, when the first request had not yet completed, it was called with a different code value.

### 2. Race condition with UI blocking fixed:
- added a new class UIBlockingProgressHUD, which synchronizes showing and hiding ProgressHUD with UI blocking;
- immediately after closing the WebView during authorization (when receiving CODE), an activity indicator is shown - UIBlockingProgressHUD is used;
- at the end of the request, regardless of the result, the activity indicator is hidden;
- the user cannot reopen the screen with the WebView while the activity indicator is on the screen.

### 3. Made a network layer for the profile screen. From a technical point of view, it consists of two network requests.
- The first request is getting basic information about the user GET /me - user login (username), first name and last name (first_name, last_name), short information (bio):
- added new service ProfileService;
- it added the fetchProfile function, which fetches the user's data via a GET /me request using the Bearer Token;
- with this data, we update the labels on the ProfileViewController screen;
- In order not to show placeholders on the profile screen, we moved getting profile data to the Splash Screen. Now we always display the data relevant at the time of launch.
- The second request is to get public information about the user GET /users/:username is the URL for getting the user's avatar (profile_image):
- Added the ProfileImageService service to get the URL of the user's avatar;
- added Observer for ProfileImageService.DidChangeNotification notification.

### 4. While the profile data is loading, a Splash Screen is displayed:
- make sure that fetchProfile is called not only after direct authorization, but also if authToken is already present at the time of launch from the viewDidAppear method.

### 5. The Kingfisher library was used to display the user's avatar.

### 6. Added SwiftKeychainWrapper to the project via SPM:
- replaced the use of OAuth2TokenStorage with storing and retrieving a token from the storage using SwiftKeychainWrapper.

### 7. When you receive an error in a network request in SplashViewController, you need to show a UIAlertController with the title “Something went wrong (), the caption “Failed to log in” and the “OK” button to hide the alert.

### 8. The key window (keyWindow) is created from the code:
- changed the application(_:, configurationForConnecting:, options:) method that creates UISceneConfiguration(name: "Default Configuration", ...), which returns the configuration created from the code (different from the standard one created from Info.plist);
- the func application(_:, configurationForConnecting:, options:) method has been changed, which now contains the creation of a rootViewController (namely, SplashViewController) from the code;
- the TabBarController class was created, which sets up its child controllers (viewControllers) in the awakeFromNib method;
- the SplashViewController class is laid out in the code, the transition to the authorization screen works correctly and does not visually differ from the version laid out in the Storyboard.

### Star Challenge ⭐️
All application screens are coded.
Elements that are currently made in the Storyboard are coded.

Dassam committed:
#11
- Changed system files: AppDelegate, SceneDelegate
- Deleted Main Storyboard
- Rewrote files so app could work from SwiftUI: AuthViewController, WebViewViewController, SplashViewController, SingleImageViewController, ImagesListCell
- Created files so app could work from SwiftUI: TabBarController
- Added network files: ProfileService, ProfileImageService, OAuth2Service
- Added data in Constants
- Added Extensions: URLSession, URLRequest, UIBlockingProgressHUD

Nikitaaeee commented on ImageFeedList/Login/ImagesList/ImagesListCell.swift:
```diff
@@ -0,0 +1,92 @@
+        self.backgroundColor = .ypBlack
+        self.selectionStyle = .none
+        
+        contentView.addSubview(mainView)
+        mainView.addSubview(cellImageView)
+        mainView.addSubview(likeButton)
+        mainView.addSubview(dateLabel)
+        
+        NSLayoutConstraint.activate([
+            mainView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 16),
+            mainView.trailingAnchor.constraint(equalTo: contentView.trailingAnchor, constant: -16),
+            mainView.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 4),
+            mainView.bottomAnchor.constraint(equalTo: contentView.bottomAnchor, constant: -4),
+            
+            cellImageView.leadingAnchor.constraint(equalTo: mainView.leadingAnchor, constant: 0),
+            cellImageView.trailingAnchor.constraint(equalTo: mainView.trailingAnchor, constant: 0),
+            cellImageView.topAnchor.constraint(equalTo: mainView.topAnchor, constant: 0),
+            cellImageView.bottomAnchor.constraint(equalTo: mainView.bottomAnchor, constant: 0),
+            
+            likeButton.trailingAnchor.constraint(equalTo: mainView.trailingAnchor, constant: 0),
+            likeButton.topAnchor.constraint(equalTo: mainView.topAnchor, constant: 0),
+            likeButton.widthAnchor.constraint(equalToConstant: 44),
+            likeButton.heightAnchor.constraint(equalToConstant: 44),
+            
+            dateLabel.leadingAnchor.constraint(equalTo: mainView.leadingAnchor, constant: 8),
+            dateLabel.bottomAnchor.constraint(equalTo: mainView.bottomAnchor, constant: -8),
+            dateLabel.widthAnchor.constraint(equalToConstant: 152),
+            dateLabel.heightAnchor.constraint(equalToConstant: 18)
+        ])
```
Мы можем визуально разгрузить инициализатор, если мы вынесем этот код в отдельные методы.

Можем создать приватный метод `addSubviews()`, в котором добавить нужные элементы, а затем создать приватный метод `applyConstraints()`, в котором добавить все нужные констрейнты.

А в этом месте уже вызывать эти методы.

Dassam replied:
Done

Nikitaaeee commented on ImageFeedList/Login/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,106 @@
+
+import UIKit
+
+final class ImagesListViewController: UIViewController {
+    private let tableView: UITableView = {
```
Круто, что не забываешь проставлять `final` и `private` везде, где это нужно! 🔥

Dassam replied:
Thx!)

Nikitaaeee commented on ImageFeedList/Login/Profile/ProfileViewController.swift:
```diff
@@ -110,12 +173,34 @@ extension ProfileViewController {
     }
 }
 
-// MARK: - Data
+// MARK: - Data Loading
 extension ProfileViewController {
     private func loadData() {
         avatarImageView.image = UIImage(named: "avatar")
         nameLabel.text = "Екатерина Новикова"
         loginNameLabel.text = "@ekaterina_nov"
         descriptionLabel.text = "Hello, world!"
     }
+    
+}
+
+
+// MARK: - Observer
```
Можешь удалять лишние пустые строки в конце методов/классов и т.д. 
Также между блоками кода принято оставлять одну пустую строку.

**SwiftLint** (сервис для чистого кода), например, может ругаться (кидать warning) на такие места,  поэтому круто на этапе обучения приучить глаза замечать такие мелочи 😊

Dassam replied:
Done

Nikitaaeee commented on ImageFeedList/Presentation/Auth/AuthViewController.swift:
```diff
@@ -36,3 +62,28 @@ extension AuthViewController: WebViewViewControllerDelegate {
+// MARK: - Layout
+
+extension AuthViewController {
+    private func configureComponents() {
+       
+        view.addSubview(logoImageView)
+        view.addSubview(loginButton)
+        
+        let safeArea = view.safeAreaLayoutGuide
+        
+        NSLayoutConstraint.activate([
+            logoImageView.widthAnchor.constraint(equalToConstant: 60),
+            logoImageView.heightAnchor.constraint(equalToConstant: 60),
+            logoImageView.centerXAnchor.constraint( equalTo: safeArea.centerXAnchor),
+            logoImageView.centerYAnchor.constraint(equalTo: safeArea.centerYAnchor),
+            
+            loginButton.centerXAnchor.constraint(equalTo: view.centerXAnchor),
+            loginButton.leadingAnchor.constraint( equalTo: safeArea.leadingAnchor, constant: 16),
+            loginButton.trailingAnchor.constraint( equalTo: safeArea.trailingAnchor, constant: -16),
+            loginButton.bottomAnchor.constraint( equalTo: safeArea.bottomAnchor, constant: -90),
+            loginButton.heightAnchor.constraint(equalToConstant: 48)
+        ])
+    }
```
Тут тоже можно разделить добавление вьюх и настройку констрейнтов в разные методы.
Тогда все будет по **SOLID** :)

Dassam replied:
Done

Nikitaaeee commented on ImageFeedList/Presentation/SingleImageView/SingleImageViewController.swift:
```diff
@@ -9,37 +9,86 @@ import UIKit
+            shareButton.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -17),
+            shareButton.widthAnchor.constraint(equalToConstant: 50),
+            shareButton.heightAnchor.constraint(equalToConstant: 50)
+        ])
+        
     }
```
Тут тоже можно убрать пустую строку

Dassam replied:
Done

Nikitaaeee commented on ImageFeedList/Presentation/SplashViewController/SplashViewController.swift:
```diff
@@ -22,24 +38,111 @@ final class SplashViewController: UIViewController {
+}
+
+/*
+
+// MARK: AuthViewController Case Transition
 extension SplashViewController {
```
Код снизу закомментирован, значит он не используется, значит он не нужен 🙂

Dassam replied:
Done

Nikitaaeee commented on ImageFeedList/Service/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,62 @@
+
+import UIKit
+
+final class ProfileImageService {
+    
+    private struct UserResult: Decodable {
+        let profileImage: ProfileImage
+    }
+    
+    private struct ProfileImage: Decodable {
+        let small: String
+        let medium: String
+        let large: String
+    }
+    
+    private var task: URLSessionTask?
+    private(set) var avatarURL: String?
+    private let urlSession = URLSession.shared
+    private var tokenStorage = OAuth2TokenStorage.shared
+        
+    static let didChangeNotification = Notification.Name(rawValue: "ProfileImageProviderDidChange")
+    static let shared = ProfileImageService()
```
Я вижу, что у тебя использован паттерн _Синглтон_ для создания _ProfileImageService_, но он используется не совсем корректно. 
Для реализации этого паттерна нам нужно: 
✅ 1. создать свойство shared, которое будет хранить в себе единый экземпляр нашего объекта: 
`static let shared = ProfileImageService() `
❌ 2. Добавить объекту-синглтону приватный инициализатор, чтобы ограничить создание новых экземпляров этого объекта:
` private init() {} `
✅ 3. Там, где мы хотим обращаться к этому объекту мы должны не создавать новый экземпляр, а обращаться к уже созданному экземпляру через свойство .shared:
` private let profileImageService = ProfileImageService.shared`

Можешь добавить приватный инициализатор, тогда мы точно будем обращаться к одному единственному экземпляру нашего класса и сможем избежать некорректного поведения при создании новых экземпляров! 
Также можешь проверить другие _Синглтоны_ в работе

Dassam replied:
У меня вообще много антипаттерна синглтона)
Подскажи пожалуйста - его ведь везде   private init() {}  надо отбивать?

Dassam replied:
Done

Nikitaaeee replied:
Да, обычно, когда мы создаем **синглтон**, первое, что нужно сделать — запретить создание объектов нашего класса извне. 
В этом нам как раз помогает **приватный** инициализатор)

Nikitaaeee commented:
Отличная работа!
Круто, что ты следишь за чистотой и аккуратностью кода.
Также молодец, что переписал все приложение через код и добавил подробное описание в пулл реквесте!
Осталось поправить **названия** UI элементов и я смогу принять работу.

Dassam committed:
#11
- Formatted code in ImagesListCell, ImagesListViewController, OAuth2TokenStorage, ProfileImageService, ProfileViewController, SingleImageViewController, SplashViewController

Dassam committed:
#11
- Refactored project after installing SwiftLint
